src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/EnumConverter.cs

```diff
@@ -439,134 +390,209 @@ internal override void WriteAsPropertyNameCore(Utf8JsonWriter writer, T value, J
                 {
-                    string name = namingPolicy.ConvertName(enumValues[i]);
-                    if (name == null)
+                    ulong fieldKey = enumField.Key;
+                    if (fieldKey == 0 ? key == 0 : (remainingBits & fieldKey) == fieldKey)
                     {
-                        ThrowHelper.ThrowInvalidOperationException_NamingPolicyReturnNull(namingPolicy);
+                        string name;
+                        if (dictionaryKeyPolicy != null && !enumField.IsAttributeOverriddenName)
+                        {
+                            name = ApplyNamingPolicy(enumField.OriginalName, dictionaryKeyPolicy);
+                            isDictionaryKeyPolicyApplied = true;
+                        }
+                        else
+                        {
+                            name = enumField.Name;
+                        }
+
+                        if (sb.Length > 0)
+                        {
+                            sb.Append(ValueSeparator);
+                        }
+
+                        sb.Append(name);
+                        remainingBits &= ~fieldKey;
+
+                        if (fieldKey == 0)
+                        {
+                            // Do not process further fields if the value equals zero.
+                            Debug.Assert(key == 0);
+                            break;
+                        }
                     }
-                    enumValues[i] = name;
                 }
 
-                converted = string.Join(ValueSeparator, enumValues);
+                if (remainingBits == 0 && sb.Length > 0)
+                {
+                    // The value is representable as a combination of flags.
+                    stringValue = sb.ToString();
+                    return true;
+                }
             }
 
-            return converted;
+            stringValue = null;
+            return false;
         }
 
-        private static string[] SplitFlagsEnum(string value)
+        private static EnumFieldInfo[] ResolveEnumFields(JsonNamingPolicy? namingPolicy, out bool containsNameAttributes)
         {
-            // todo: optimize implementation here by leveraging https://github.com/dotnet/runtime/issues/934.
-            return value.Split(
+            containsNameAttributes = false;
 #if NET
-                ValueSeparator
+            string[] names = Enum.GetNames<T>();
+            T[] values = Enum.GetValues<T>();
 #else
-                new string[] { ValueSeparator }, StringSplitOptions.None
+            string[] names = Enum.GetNames(typeof(T));
+            T[] values = (T[])Enum.GetValues(typeof(T));
 #endif
-                );
+            Debug.Assert(names.Length == values.Length);
+
+            Dictionary<string, string>? enumMemberAttributes = null;
+            foreach (FieldInfo field in typeof(T).GetFields(BindingFlags.Public | BindingFlags.Static))
```

Comment:
cc @eerhardt @vitek-karas

This feature uses reflection to read attributes off enum members in both reflection and source gen. I've added a trimming test that should validate that this should work as expected.

Reply:
How come you're not getting trim warnings? `T` is not annotated...

@sbomer, @MichalStrehovsky for trim/AOT side of things.

Reply:
Oh - I didn't see https://github.com/dotnet/runtime/pull/100814... nice.

Reply:
My initial prototype was adding viral DynamicallyAccessedMember annotations everywhere, so I was very surprised when I deleted and everything kept working. That should explain it.